ember-headless-form/src/components/headless-form.ts

```diff
@@ -146,4 +219,36 @@ export default class HeadlessFormComponent<
   set<KEY extends FormKey<FormData<DATA>>>(key: KEY, value: DATA[KEY]): void {
     this.internalData[key] = value;
   }
+
+  @action
+  async handleFieldValidation(e: Event): Promise<void> {
+    const { target } = e;
+    const { name } = target as HTMLInputElement;
+
+    if (name) {
+      const field = this.fields.get(name as FormKey<FormData<DATA>>);
+
+      if (field) {
+        this.lastValidationResult = await this.validate();
+        field.validationEnabled = true;
+      }
+    } else {
+      // @todo how to handle custom controls that don't emit focusout/change events from native form controls?
+    }
+  }
+
+  @action
+  async handleFieldRevalidation(e: Event): Promise<void> {
+    const { target } = e;
+    const { name } = target as HTMLInputElement;
+
+    if (name) {
+      if (this.showErrorsFor(name as FormKey<FormData<DATA>>)) {
+        this.lastValidationResult = await this.validate();
```

Comment:
interacting with `this` after an await isn't safe 🙈 

I keep meaning to work on [this lint](https://github.com/ember-cli/eslint-plugin-ember/pull/1421) but... priorities 🙃 

be sure to `if (isDestroyed(this) || isDestroying(this)) return` after an await

Reply:
Good catch! So I can certainly add `isDestroy{ing,ed}` guards here, even if it's just for code style or best practices.

But I'd first like to understand why this could be a problem here, i.e. what is unsafe about it? 

So when the async action still has a pointer to `this` while the component that `this` refers to is already "destroyed" (unmounted), then `this` will still exist. There is no way it could be "gone", because that's how Garbage Collection works: here our async action is basically the retainer for that object preventing GC.

So it exists, but it could be in a "destroyed" state. In classic Ember, a call to `set(this,...)` or `this.set()` would fail, so yes, that would be bad. But in Octane, would mutating a tracked property be so bad? I don't think anything will throw here, right? And when our form is already "unmounted" (in React's terminology, do we have a better term, unrendered?), this tracked property will just not be consumed anymore, so also not cause rerenders, but that should be all, no further harm caused, or am I missing something?

Compare that to [this comment](https://github.com/ember-cli/eslint-plugin-ember/issues/418#issuecomment-1009181665) here on the linting issue...

Loosely related: should we wrap the async methods (that call potentially async validators) with `waitFor()` from `@ember/test-waiters`, so there are able to cover async things happening that Ember does not auto-detect as async? I would think so, yes?

Reply:
>  I would think so, yes?

yeah, that's a good way to solve this whole class of problem.

> what is unsafe about it?

you raise some good points, and I'm going to have to do some concrete research with shareable examples before continuing on with the lint 😅

Reply:
> accessing services that is the problem in this situation, since service instantiation is lazy

Oh right, that can be problematic indeed!

Reply:
So I added `@waitFor` to cover async validations, and made the test cover async validations properly (i.e. they were indeed failing without that `@waitFor`), see [this commit](https://github.com/CrowdStrike/ember-headless-form/pull/25/commits/0fb2cec919c1b5a85124c6bb01e2e9e3d89dae5e)

Regarding the `isDestroy{ing,ed}` topic, should we address this here and now, or eventually iterate on it when it's more clear how to address this?

In other words: is this good to merge? @NullVoxPopuli @nicolechung

Reply:
we're good to merge. 

isDetroy{ing,ed}, I think I want to iterate on the lint, so we can have a definite error case to match on, and then provide an autofix as well